src/streams/rtx_cache_buf.rs

```diff
@@ -0,0 +1,383 @@
+use std::mem;
+use std::time::{Duration, Instant};
+
+use crate::util::already_happened;
+
+/// Fixed size growing buffer that evicts the oldest entries based on time.
+///
+/// The buffer should evict entries based on time, not max size. If we
+/// detect an entry overwriting a current entry, we will grow the buffer
+/// on next eviction run.
+#[derive(Debug)]
+pub struct EvictingBuffer<T> {
+    buf: Vec<Option<Entry<T>>>,
+    /// How long to keep entries for.
+    max_age: Duration,
+    /// Last inserted position
+    last_position: Option<u64>,
+    /// Next element to evict.
+    next_evict: Option<u64>,
+    // Number of times we overwritten the tail instead of an evicted value since the last timeout.
+    overwrites: u16,
+    // Last timeout when we evicted elements.
+    last_timeout: Instant,
+}
+
+/// Entry in the buffer to keep track of position and time separately.
+#[derive(Debug)]
+struct Entry<T> {
+    /// Position is offset modulus the buffer size.
+    position: u64,
+    /// Entry time. Used for eviction.
+    timestamp: Instant,
+    /// The value held at this entry.
+    value: T,
+}
+
+// We don't want to require that T is Clone/Copy, which means we
+// must do this instead of using the vec![] macro.
+fn prepare_buf<T>(len: usize) -> Vec<Option<T>> {
+    let mut buf = Vec::with_capacity(len);
+    for _ in 0..len {
+        buf.push(None);
+    }
+    buf
+}
+
+impl<T> EvictingBuffer<T> {
+    /// Creates a new buffer with an initial size.
+    pub fn new(initial_size: usize, max_age: Duration) -> Self {
+        Self {
+            buf: prepare_buf(initial_size),
+            max_age,
+            last_position: None,
+            next_evict: None,
+            overwrites: 0,
+            last_timeout: already_happened(),
+        }
+    }
+
+    fn index_for_position(&self, position: u64) -> usize {
+        (position % self.buf.len() as u64) as usize
+    }
+
+    /// Push a new entry.
+    ///
+    /// Position is an increasing sequence number. The sequence can be out of order.
+    pub fn push(&mut self, position: u64, timestamp: Instant, value: T) {
+        if self.buf.is_empty() || self.max_age.is_zero() {
+            return;
+        }
+
+        if timestamp < self.last_timeout {
+            // Value is already considered evicted.
+            return;
+        }
+
+        // First element.
+        let next_evict = if let Some(v) = self.next_evict {
+            v
+        } else {
+            self.next_evict = Some(position);
+            position
+        };
+
+        if position < next_evict {
+            // Do not cache values older than evict position.
+            return;
+        }
+
+        let index = self.index_for_position(position);
+
+        if let Some(entry) = &self.buf[index] {
+            // If the position is exactly the same, we for some reason are overwriting the
```

Comment:
When we cache the first entry, `next_evict` is == `position`. Is there another case for this situation ? Because if there is, we risk overwriting `self.next_evict` with new data, which blocks eviction around line :172 I think

Reply:
True. We should probably move next_evict one position up in this case.

Reply:
I can make a test for this.

Reply:
It is a problem. I fixed it.